feat: `avm/ptn/data/private-analytical-workspace` V0.1

jbinko opened the pull request:
## Description

<!--
>Thank you for your contribution !
> Please include a summary of the change and which issue is fixed.
> Please also include the context.
> List any dependencies that are required for this change.

Fixes #123
Fixes #456
Closes #123
Closes #456

For badge copy from the build action, three dots icon and "create status badge"
-->

This is the initial implementation of the proposed pattern module detailed in proposal. <https://github.com/Azure/Azure-Verified-Modules/issues/949> It includes documentation, unit tests, and PowerShell post-unit tests.

## Pipeline Reference

<!-- Insert your Pipeline Status Badge below -->

| Pipeline |
| -------- |
[![avm.ptn.data.private-analytical-workspace](https://github.com/jbinko/bicep-registry-modules/actions/workflows/avm.ptn.data.private-analytical-workspace.yml/badge.svg?branch=private-analytical-workspace&event=workflow_dispatch)](https://github.com/jbinko/bicep-registry-modules/actions/workflows/avm.ptn.data.private-analytical-workspace.yml)

## Type of Change

<!-- Use the checkboxes [x] on the options that are relevant. -->

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [x] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [x] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [x] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

<!--  Please keep up to date with the contribution guide at https://aka.ms/avm/contribute/bicep -->

AlexanderSehr commented on .github/workflows/avm.ptn.data.private-analytical-workspace.yml:
```diff
@@ -0,0 +1,91 @@
+name: "avm.ptn.data.private-analytical-workspace"
+
+on:
+  schedule:
+    - cron: "0 12 1/15 * *" # Bi-Weekly Test (on 1st & 15th of month)
```
```suggestion
```

AlexanderSehr commented on .github/workflows/avm.ptn.data.private-analytical-workspace.yml:
```diff
@@ -0,0 +1,91 @@
+        type: string
+        description: "Default location overwrite (e.g., eastus)"
+        required: false
+        default: 'westeurope'
```
```suggestion
```

If you set the default here it would be the single location used for all tests instead of using the location rotation.

AlexanderSehr replied:
Hey @jbinko,
how come this comment was resolved? The line is still in the file. 
Also (unrelated to the issue at hand), I'd recommend to only resolve comments that are auto-resolved by GitHub (e.g., suggestions) and only comment on all other comments. This makes a re-review a lot easier :)

AlexanderSehr commented on .github/workflows/avm.ptn.data.private-analytical-workspace.yml:
```diff
@@ -0,0 +1,91 @@
+        default: 'westeurope'
+  push:
+    branches:
+      - private-analytical-workspace
```
```suggestion
      - main
```

AlexanderSehr commented on avm/ptn/data/private-analytical-workspace/tests/common.tests.ps1:
```diff
@@ -0,0 +1,393 @@
+function Test-VerifyLock($ResourceId) {
```
I'm not sure this file will be called by the CI in its current location. If you want post-deployment tests, they should be in the same folder (or sub-folder) of the test the post-deployment test should run for

AlexanderSehr commented on avm/ptn/data/private-analytical-workspace/tests/e2e/defaults/main.test.bicep:
```diff
@@ -0,0 +1,73 @@
+    scope: resourceGroup
+    name: '${uniqueString(deployment().name, resourceLocation)}-test-${serviceShort}-${iteration}'
+    params: {
+      // You parameters go here
+      name: '${namePrefix}${serviceShort}001'
+      tags: {
+        Owner: 'Contoso'
+        CostCenter: '123-456-789'
+      }
```
```suggestion
      name: '${namePrefix}${serviceShort}001'
```

The defaults test should only contain required parameters

AlexanderSehr commented on avm/ptn/data/private-analytical-workspace/tests/e2e/max/main.test.bicep:
```diff
@@ -0,0 +1,96 @@
+    scope: resourceGroup
+    name: '${uniqueString(deployment().name, resourceLocation)}-test-${serviceShort}-${iteration}'
+    params: {
+      // You parameters go here
```
```suggestion
```

AlexanderSehr commented on avm/ptn/data/private-analytical-workspace/tests/e2e/max/main.test.bicep:
```diff
@@ -0,0 +1,96 @@
+      // You parameters go here
+      name: '${namePrefix}${serviceShort}001'
+      location: resourceLocation
+      //solutionAdministrators: [
+      //  {
+      //    principalId: '<EntraGroupId>'
+      //    principalType: 'Group'
+      //  }
+      //]
+      tags: {
+        Owner: 'Contoso MAX Team'
+        CostCenter: '123459876'
+      }
+      enableTelemetry: true
+      enableDatabricks: true
+      //virtualNetworkResourceId: null
+      //logAnalyticsWorkspaceResourceId: null
+      //keyVaultResourceId: null
```
Are the comments leftovers?
```suggestion
      tags: {
        Owner: 'Contoso MAX Team'
        CostCenter: '123459876'
      }
      enableDatabricks: true
```

AlexanderSehr commented on avm/ptn/data/private-analytical-workspace/tests/e2e/min-priv/main.test.bicep:
```diff
@@ -0,0 +1,86 @@
+    scope: resourceGroup
+    name: '${uniqueString(deployment().name, resourceLocation)}-test-${serviceShort}-${iteration}'
+    params: {
+      // You parameters go here
```
I take it the comments are leftovers?

jbinko replied:
Idea was to document what is available. Removing Comments.

AlexanderSehr commented on avm/ptn/data/private-analytical-workspace/tests/e2e/min-pub/main.test.bicep:
```diff
@@ -0,0 +1,86 @@
+    scope: resourceGroup
+    name: '${uniqueString(deployment().name, resourceLocation)}-test-${serviceShort}-${iteration}'
+    params: {
+      // You parameters go here
```
Leftover comments?

jbinko replied:
Idea was to document what is available. Removing Comments.

AlexanderSehr commented on avm/ptn/data/private-analytical-workspace/tests/e2e/uc01-priv/main.test.bicep:
```diff
@@ -0,0 +1,86 @@
+    scope: resourceGroup
+    name: '${uniqueString(deployment().name, resourceLocation)}-test-${serviceShort}-${iteration}'
+    params: {
+      // You parameters go here
```
Leftover comments?

jbinko replied:
Idea was to document what is available. Removing Comments.

AlexanderSehr commented on avm/ptn/data/private-analytical-workspace/tests/e2e/uc01-pub/main.test.bicep:
```diff
@@ -0,0 +1,86 @@
+    scope: resourceGroup
+    name: '${uniqueString(deployment().name, resourceLocation)}-test-${serviceShort}-${iteration}'
+    params: {
+      // You parameters go here
```
Leftover comments?

jbinko replied:
Idea was to document what is available. Removing Comments.

AlexanderSehr commented on avm/ptn/data/private-analytical-workspace/tests/e2e/uc02-priv/dependencies.bicep:
```diff
@@ -0,0 +1,223 @@
+var subnets = [
+  {
+    name: subnetName01
+    addressPrefix: '10.0.0.0/24'
```
To make it easier to maintain I'd recommend to move the vnet's address space to a variable somewhere on top (below the parameters) and then use the CIDR function to calculate the address spaces. You can find an example of how the function is used in many other tests.

AlexanderSehr replied:
The same comment applies to the other test files

jbinko replied:
Idea was to document what is available. Removing Comments.

jbinko replied:
I like it. Adopted!

AlexanderSehr commented on avm/ptn/data/private-analytical-workspace/tests/e2e/uc02-priv/main.test.bicep:
```diff
@@ -0,0 +1,114 @@
+// Dependencies //
+// ============ //
+
+var subnetName01 = 'private-link-subnet'
+var subnetName02 = 'dbw-frontend-subnet'
+var subnetName03 = 'dbw-backend-subnet'
```
if you want to align this to other tests, I'd recommend to hardcode these values in the subnets of the dependencies.bicep file directly as they cannot conflict with other resources. This possiblity of a conflict is actualy the main reason why we ask contributors to generate all top-level resource names in this file and pass it into the dependencies.bicep file. However, for anything that is then internal to those resources we don't require it anymore and it frankly makes the file easier if minimize the amount of parmaeters and keep the sub-resource names close to where they are used 😉 
No requirement, but a recommendation.

jbinko replied:
Yes, make sense. Refactored dependencies.bicep files.

AlexanderSehr replied:
Hey @jbinko,
thanks for that. Unfortunately we went a 'bit' too far in the opposite direction now 😄 Let me elaborate on my initial comment a bit more:
What we usually suggest is to have the names of all top-level resources (vnet, key vault, law, you name it) in the main.test.bicep, where the name is created and the values are passed into the `dependencies.bicep`. For example like here: 
https://github.com/Azure/bicep-registry-modules/blob/b673ec8840fa9eee5e71fa76bf991b0899d5714d/avm/res/key-vault/vault/tests/e2e/max/main.test.bicep#L34-L42

In the above example, you can also see the main reason we do that: We inject the `namePrefix` & `serviceShort` in each resource name + we add a `dep-` prefix to all resources. 
So you may wonder why:
- The `namePrefix` & `serviceShort` are here because their combination is always different from fork to fork and test case to test case. So there won't be any conflicts with other contributors, no you without yourself when running multiple pipelines
- Also, the `dep-` prefix ensures that our WAF-aligned tests (via PSRule) ignore these resources when testing. Reason being: We want PSRule to test you module, not your module's temporary test dependencies.
- (Bonus): Because the names of the top-level resources are generated in the `main.test.bicep` file it's a 'bit' easier for troubleshooting as you can immediately see where a name an error message may contain comes from. Not super important, but a nice side-effect.

Now the one thing we don't have here are the nested resources. Like the subnet names I called out, or let's say the names of secrets in a key vault. As those are 'inner' resources to a 'main' resource, they usually never conflict with anything. These are hence the values we expect contributors to hardcode in their `dependencies.bicep` files als none of the previously mentioned benefits apply for them.

All in all, it keeps both your `main.test.bicep` & `depepdencies.bicep` file(s) quite distinct in scope and easy to understand (or that's at least the idea 😄). I hope the above made sense.

AlexanderSehr commented on avm/ptn/data/private-analytical-workspace/tests/e2e/uc02-priv/main.test.bicep:
```diff
@@ -0,0 +1,114 @@
+    scope: resourceGroup
+    name: '${uniqueString(deployment().name, resourceLocation)}-test-${serviceShort}-${iteration}'
+    params: {
+      // You parameters go here
```
Leftover?

jbinko replied:
Idea was to document what is available. Removing Comments.

AlexanderSehr commented on avm/ptn/data/private-analytical-workspace/tests/e2e/uc02-pub/main.test.bicep:
```diff
@@ -0,0 +1,114 @@
+    scope: resourceGroup
+    name: '${uniqueString(deployment().name, resourceLocation)}-test-${serviceShort}-${iteration}'
+    params: {
+      // You parameters go here
```
Leftover?

jbinko replied:
Idea was to document what is available. Removing Comments.

AlexanderSehr commented on avm/ptn/data/private-analytical-workspace/tests/e2e/uc03-priv/main.test.bicep:
```diff
@@ -0,0 +1,116 @@
+    scope: resourceGroup
+    name: '${uniqueString(deployment().name, resourceLocation)}-test-${serviceShort}-${iteration}'
+    params: {
+      // You parameters go here
```
Leftover?

jbinko replied:
Idea was to document what is available. Removing Comments.

AlexanderSehr commented on avm/ptn/data/private-analytical-workspace/tests/e2e/uc03-pub/main.test.bicep:
```diff
@@ -0,0 +1,116 @@
+    scope: resourceGroup
+    name: '${uniqueString(deployment().name, resourceLocation)}-test-${serviceShort}-${iteration}'
+    params: {
+      // You parameters go here
```
Leftover comments?

jbinko replied:
Idea was to document what is available. Removing Comments.

AlexanderSehr commented on avm/ptn/data/private-analytical-workspace/tests/e2e/waf-aligned/main.test.bicep:
```diff
@@ -0,0 +1,98 @@
+    scope: resourceGroup
+    name: '${uniqueString(deployment().name, resourceLocation)}-test-${serviceShort}-${iteration}'
+    params: {
+      // You parameters go here
```
Leftover comments?

jbinko replied:
Idea was to document what is available. Removing Comments.

jbinko commented:
> @jbinko We have added examples for Bicep parameter files to the Readme. This has been applied to all published modules but needs to be done for PRs as well. Can you please update your branch and run the Set-AVMModule utility as detailed [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/generate-bicep-module-files/). It is required for the validation pipeline to succeed and the contribution to be published.
> 
> Please reach out if any support is needed.

Updated documentation

jbinko commented:
All review comments are resolved.
Ready for next review.

BTW - I cannot run pipeline and execute tests. Something is wrong at the step "Set environment"
E: Unable to locate package powershell

It looks like generic issue. I'm seeing this for other pipelines as well.

AlexanderSehr commented:
> All review comments are resolved. Ready for next review.
> 
> BTW - I cannot run pipeline and execute tests. Something is wrong at the step "Set environment" E: Unable to locate package powershell
> 
> It looks like generic issue. I'm seeing this for other pipelines as well.

Oh wow. Can't say I've seen that before. 👀 
Any ideas @eriqua? If it does not work at all, I guess we could merge the PR into an upstream branch once it is ready to validate it there. But let's only go there if there is no other way as it would be much better to find out 'where' this issue comes from also for future contibutions.

AlexanderSehr commented on avm/ptn/data/private-analytical-workspace/main.bicep:
```diff
@@ -0,0 +1,864 @@
+// Definitions      //
+// ================ //
+
+type lockType = {
```
We'll soon at a spec to ask contributors to add the `@export()` tag to their UDTs - if - they should not remain private. This has the perk that consumers of your module can import those types in their templates when using your module to re-use them. 
If you don't mind, I'd ask you to please add this small annotation to each of your types. There is no impact on your module's functionality.

jbinko replied:
Implemented propose changes

AlexanderSehr commented:
> All review comments are resolved. Ready for next review.
> 
> BTW - I cannot run pipeline and execute tests. Something is wrong at the step "Set environment" E: Unable to locate package powershell
> 
> It looks like generic issue. I'm seeing this for other pipelines as well.

3 comments left 💪 Thanks @jbinko

eriqua commented:
> > All review comments are resolved. Ready for next review.
> > BTW - I cannot run pipeline and execute tests. Something is wrong at the step "Set environment" E: Unable to locate package powershell
> > It looks like generic issue. I'm seeing this for other pipelines as well.
> 
> Oh wow. Can't say I've seen that before. 👀 Any ideas @eriqua? If it does not work at all, I guess we could merge the PR into an upstream branch once it is ready to validate it there. But let's only go there if there is no other way as it would be much better to find out 'where' this issue comes from also for future contibutions.

@AlexanderSehr only as a transient error unfortunately. From the history of @jbinko's fork this seems pretty consistent.
@jbinko did you get any other pipeline to pass the set-environment point?

AlexanderSehr commented:
Small addition: It seems to be related to: https://github.com/PowerShell/PowerShell/issues/23197#issuecomment-2402934692
Currently testing if we still need the explicit pwsh version installation (which was added due to a bug of a version). If we do, then we need to hardcode the runner version

jbinko commented:
OK, all changes completed on my side. Ready for merge. Tests and deployments are passing. @AlexanderSehr pipeline succeeds if I do changes  you proposed in the file \avm\utilities\pipelines\sharedScripts\Set-EnvironmentOnAgent.ps1. Those changes are not included in this PR. I removed them after try. I expect this will be changed in the main in original repo.

AlexanderSehr commented:
Hey @jbinko,
there are 2 final comments hidden behind this GitHub feature 😄 
![image](https://github.com/user-attachments/assets/62bbde34-7c7c-45ce-8278-9ec67b387865)

jbinko commented:
> Hey @jbinko, there are 2 final comments hidden behind this GitHub feature 😄 ![image](https://private-user-images.githubusercontent.com/5365358/376592636-62bbde34-7c7c-45ce-8278-9ec67b387865.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkxNTg0MzIsIm5iZiI6MTcyOTE1ODEzMiwicGF0aCI6Ii81MzY1MzU4LzM3NjU5MjYzNi02MmJiZGUzNC03YzdjLTQ1Y2UtODI3OC05ZWM2N2IzODc4NjUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTdUMDk0MjEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWMwYWIzZmVlMTQ3N2M2MDcwMDNjOTIwYzkxMGQ2ZDk4NjFhNjJiYTYyZWVjNDUyYTdkMmNlOTA3YTk2MzRhMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Fnlof9dP7VtqHU3nPJbdWqkN_p2XKCv41Nfnp_RPDNg)

Thanks @AlexanderSehr - got it. I fixed the remove of default test location and virtual network name propagation into dependencies files in tests.

Should be ready for merge now.

AlexanderSehr commented:
Hey @jbinko,
please not that many of the module's post-deployment tests did fail in the upstream workflow ([ref](https://github.com/Azure/bicep-registry-modules/actions/runs/11427911716)). That means the module was **not** published. Could you please check if you can reproduce the errors you can see in the workflow's log? The deployment seemed to have worked.

jbinko commented:
> Hey @jbinko, please not that many of the module's post-deployment tests did fail in the upstream workflow ([ref](https://github.com/Azure/bicep-registry-modules/actions/runs/11427911716)). That means the module was **not** published. Could you please check if you can reproduce the errors you can see in the workflow's log? The deployment seemed to have worked.

Thanks @AlexanderSehr, I created new PR #3626 which addresses the post deployment tests issue.
The tests are run by default in the fixed location North Europe
